### PR TITLE
Autogenerate Version on build

### LINKED ISSUE #91: Automatically generated version on build
I think we can learn Golang's version management. 

On master, all the version numbers will be gave on build and the version number is the commit hash. Since the hash is too long, we can use the first 7 characters.  i.e. tip+2b445c7

On release branch, all the version number will be gave on build and the version number is branch name plus git hash. i.e. 1.0+2b445c7

For a tag version i.e. 1.0.0, the version number could be static and we can give it a special number not a git hash.

So that, when a user report an issue, we can know the question is from which version quickly.

### PR DESCRIPTION
**Pull request by bkcsoft**
Fixes #91 
Closes #101

### CODE REVIEW

**Comment by appleboy** [comment]
ＬＧＴＭ

**Comment by bkcsoft** [comment]
@appleboy lgtm only works if it's in English :wink:

**Comment by 0xbaadf00d** [comment]
Is the file `templates/.VERSION` still necessary in the Git repo ?

File: templates/.VERSION
Comment (thibaultmeyer) [inline templates/.VERSION]: Is the file `templates/.VERSION` still necessary in the Git repo now it was auto-generated ? Maybe remove it and update `.gitignore` make sense ?

Comment (strk) [reply templates/.VERSION]: I think the .VERSION file is there as a kind of "safety measure" to avoid using a set ot templates meant for a different version of the built gitea. If we want to obtain the same safety we should still include a version but maybe keep it separated from the application version (a template API version). Could easily become hard to maintain though.

Or, we give up on the attempt and just print a WARNING whenever templates/ are not embedded (see #tboerger PR about embedding template/ and public/)

Comment (0xbaadf00d) [inline templates/.VERSION]: Currently, if I clone and dev feature, Gitea will never run because i'm not necessary check out at commit g2c92e45.

**Comment by strk** [comment]
On Thu, Nov 17, 2016 at 12:10:44PM -0800, Thibault Meyer wrote:

> Currently, if I clone and dev feature, Gitea will never run because i'm not necessary check out at commit g2c92e45.

Good observation. I suggest the check is only made on
Major.Minor

--strk;

File: Makefile
Comment (strk) [inline Makefile]: As the file is retained in the repo, it should not be cleaned by "make clean"

File: Makefile
Comment (tboerger) [inline Makefile]: IMHO we should drop this version file. People that build from source should update the templates on their own.

File: main.go
Comment (tboerger) [inline main.go]: Let's do it without the v

File: Makefile
Comment (tboerger) [inline Makefile]: This covers only tags and nothing else, beside that it doesn't provide the commit sha

**Review by tboerger — CHANGES_REQUESTED** [review]

**Comment by strk** [comment]
On Thu, Nov 17, 2016 at 01:06:10PM -0800, Thomas Boerger wrote:

> +VERSION = $(shell git describe --tags --always)
> 
> This covers only tags and nothing else, beside that it doesn't provide the commit sha 

The output of that command actually includes the closest tag label,
the count of commits since that tag _and_ commit sha

File: templates/.VERSION
Comment (bkcsoft) [reply templates/.VERSION]: @0xbaadf00d removal of `templates/.VERSION` is done by @tboerger in #74 https://github.com/go-gitea/gitea/pull/74/files#diff-1101bb4d7aabf48ba4f4b98ecd7f3efb

File: Makefile
Comment (bkcsoft) [reply Makefile]: It has to be because `make` :unamused:

File: Makefile
Comment (bkcsoft) [reply Makefile]: nope

```
$ git describe --tags --always
v0.9.99-213-gec33e9c
```

File: main.go
Comment (bkcsoft) [reply main.go]: that's what `git describe` spits out :roll_eyes:

**Comment by bkcsoft** [comment]
@strk 

> I suggest the check is only made on Major.Minor

I agree, but suggest we remove everything but the tag instead, e.g. `v1.0.0`

File: main.go
Comment (bkcsoft) [reply main.go]: technically, I can just do `var Version string` and be done with it. It _should_ be set by `make build` anyhow

File: Makefile
Comment (bkcsoft) [reply Makefile]: I was leaving that to your PR (#74) so you wouldn't get conflicts :stuck_out_tongue:

File: Makefile
Comment (bkcsoft) [reply Makefile]: Maybe even slap `--dirty` on it, so **if** you have a dirty tree you get `<tag>-dirty` instead

```
$ git describe --tags --dirty
v0.9.99-213-gec33e9c-dirty
```

**Comment by lunny** [comment]
@appleboy only members of maintainer team could make an available LGTM. Please see `CONTRIBUTING.md` for how to apply as a maintainer.

File: main.go
Comment (lunny) [inline main.go:21]: I think `Var Version = "0.9.0+dev"` is better. So that if user `go get` then the version is shown it's a dev version.

File: templates/.VERSION
Comment (lunny) [inline templates/.VERSION]: The version file also could be `0.9.0+dev` so that `go get` will not report version error.

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
I think we should always keep valid semver versions

```
go build on branch master: 1.0.0+dev, 1.1.0+dev, 1.2.0+dev
go build on branch v1.0: 1.0.0+dev, 1.0.1+dev, 1.0.2+dev
go build on tag v1.0.0: 1.0.0+dev, 1.0.1+dev, 1.0.2+dev

make on branch master: 1.0.0+631c18a, 1.1.0+631c18a, 1.2.0+631c18a
make on branch v1.0: 1.0.0+631c18a, 1.0.1+631c18b, 1.0.2+631c18c
make on tag v1.0.0: 1.0.0
```

**Comment by appleboy** [comment]
@lunny I got it. thanks.

**Comment by bkcsoft** [comment]
@lunny These are valid semvers :slightly_smiling_face: 
Paste `v0.9.99-213-gec33e9c` into http://jubianchi.github.io/semver-check/ and check :)

**Comment by bkcsoft** [comment]
@tboerger @lunny please re-review :)

File: main.go
Comment (strk) [reply main.go:21]: What I do in other projects is always set the target version for each branch (ie: `<next version>-alpha`) and then separately and optionally append a changeset identifier which is generated on "make". As `Go` doesn't let us autogenerate things such setup would result in `<next version>-alpha` resulting in `go` obtained builds and `<next version>-alpha+<metadata>` resulting from builds via `make`.

File: main.go
Comment (tboerger) [reply main.go]: But than go get doesn't provide any version. And just because git spits that out is not a reason to blindly take it.

**Comment by strk** [comment]
On Thu, Nov 17, 2016 at 05:57:07PM -0800, bkcsoft wrote:

> @lunny These are valid semvers :slightly_smiling_face: 
> Paste `v0.9.99-213-gec33e9c` into http://jubianchi.github.io/semver-check/ and check :)

What's not valid is that it is considered to come _before_ rather than
_after_ 0.9.99.

ie: The next patch release will be 0.9.99

But I'm not sure that checker is really implementing the specs
correctly, as it thinks `> v0.9.99` is not true for `v1.0.0-alpha1`

File: templates/.VERSION
Comment (strk) [inline templates/.VERSION]: It should probably be `v0.9.99<something>`, so to be `> v0.9.9', which is the closest tag to the`master` branch

Comment (lunny) [reply templates/.VERSION]: No. Please see my document. @strk

Comment (strk) [reply templates/.VERSION]: @lunny what I'm saying is not related to semver format but to reality. We tagged `v0.9.99`, so current template .VERSION cannot be in the past, and `v0.9.0+dev` is in the past, no ?

**Comment by lunny** [comment]
I tag `1.x.x`. When I `make build`, the version is `1.x.x+7-g0a8f272`?
I think that's not what we expected.
If we make a tag, the version should be only the `1.x.x`, not others.
If we currently not make a tag, the version should be `1.x.x+0a8f272`

**Comment by strk** [comment]
Upon checking out a tagged hash, you get only the tag from `git describe`:

```
$ git co v0.9.99
Note: checking out 'v0.9.99'.
$ git describe --tag --always
v0.9.99
```

File: main.go
Comment (tboerger) [inline main.go]: It's 1.0.0 :)

Comment (tboerger) [reply main.go]: +dev

File: templates/.VERSION
Comment (tboerger) [reply templates/.VERSION]: >= 1.0.0+dev?

**Review by tboerger — CHANGES_REQUESTED** [review]
The version should not include the v, so let's drop it from the go files and from the git tag.

File: main.go
Comment (lunny) [reply main.go]: `1.0.0+dev` means we have released 1.0 and begin dev.
`0.9.0+dev` means we have released 0.9 and begin dev.

Comment (strk) [reply main.go]: `v0.9.9` is tagged, so `0.9.0+dev` is still too low, should at least be `0.9.9+dev`
Or, if we hard-code something, we could hard code the *target* and make it `1.0.0-alpha.<numcommits-since-last-tag>+<sha>`

Comment (strk) [reply main.go]: (or just `1.0.0-alpha` as we cannot really hard-code `sha`)

Comment (strk) [reply main.go]: I guess `1.0.0-dev` would have the same effect, btw (being *before* `1.0.0`)

**Comment by bkcsoft** [comment]
@tboerger The previous tag `v0.9.97` included `v`, hence why it stuck. We could change that, though having `v` in the beginning is basically de-facto standard and most (if not all) SemVer-libs silently drops the initial `v` so I don't really see the point in dropping it.

**Comment by strk** [comment]
On Mon, Nov 21, 2016 at 04:34:41AM -0800, bkcsoft wrote:

> @strk `v0.9+dev` would work?

v0.9+dev seems to be considered invalid by
http://jubianchi.github.io/semver-check/

Probably due to missing Patch level number

File: main.go
Comment (tboerger) [reply main.go]: It is the upcoming 1.0.0, it doesn't mean we have released it.

**Comment by tboerger** [comment]
For tags it's common to include the v, but for the version string itself it's uncommon, that's why I would keep it for the tag but drop it from the version string

File: main.go
Comment (lunny) [reply main.go]: so `1.0.0-dev`

**Comment by bkcsoft** [comment]
> I would keep it for the tag but drop it from the version string  

@tboerger Which would require _yet_ another sed-line, which is irritating 😒 

`git describe --tags --always | sed 's/-/+/'` => `git describe --tags --always | sed 's/-/+/' | sed 's/^v//'`

**Comment by bkcsoft** [comment]
@tboerger @lunny done, it's now `1.0.0+dev` for the next release :)

**Comment by strk** [comment]
Should be `1.0.0-dev`, not `+dev`, as we want it to be evaluated as being *less* than `1.0.0`

**Comment by bkcsoft** [comment]
@strk no we want it to evaluate as `=1.0.0` since it's for the upcoming release (see above comments by lunny and tboerger), and since `dev` isn't part of the version but rather just metadata, it should be `+dev` according to SemVer 🙂 

I don't agree with setting the version in source before tagging it, but it's 2 against 1 here so I'm out of luck

**Comment by strk** [comment]
Ok, if it needs be `=1.0.0` then `1.0.0+dev` is fine

**Comment by tboerger** [comment]
If you are unsure check deb or rpm version guidelines

**Review by tboerger — APPROVED** [review]

**Comment by tboerger** [comment]
LGTM

**Comment by lunny** [comment]
```
[....io/gitea/cmd/web.go:86 checkVersion()] [E] Binary version is lower than template file version, did you forget to recompile Gogs?
```

**Comment by lunny** [comment]
I think maybe you have to write the version to .VERSION when `make build`.

**Comment by lunny** [comment]
But the makefile will generate binary to `./bin` and the `templates` dir is on root dir. @tboerger

**Comment by bkcsoft** [comment]
@lunny Fix that, new-lines in `templates/.VERSION` is discouraged 😆

**Comment by bkcsoft** [comment]
@strk Open an issue :trollface:

**Comment by lunny** [comment]
@bkcsoft @tboerger for this PR, it seems it works. But for a `make build`, it's not ready to work. Just build a binary to `./bin` dir and have no templates or statics copied and the `.VERSION` file doesn't be write the version.

**Comment by tboerger** [comment]
We can also put the binary from make build into the root, I have already said that elsewhere

**Comment by lunny** [comment]
OK. We can send another PR to change build target from `./bin` to root dir. But since we changed the main.Version to 1.0.0 but `templates/.VERSION` is still 1.0.0+dev. `Gitea web` will be failed becuase that.

**Comment by tboerger** [comment]
But the main version is still 1.0.0+dev

**Comment by lunny** [comment]
So I think we should write the generated verson to .VERSION in makefile on this PR. @bkcsoft

**Comment by bkcsoft** [comment]
@lunny `make build` already updates `.VERSION` with the correct version...
@tboerger it's set to `1.0.0+dev` 😕

**Comment by tboerger** [comment]
That's what I mean :)

**Comment by bkcsoft** [comment]
When someone merges this, please **Squash Merge** it. Don't need 8 commits in the history for this simple fix 😆 

sm_github

File: Makefile
Comment (strk) [inline Makefile:12]: nitpicking, you should be able to group the sed calls: `sed 's/^v//;s/-/+/'`

Comment (bkcsoft) [reply Makefile:12]: @strk that depends entirely on the version of `sed` unfortunately... FreeBSD for example doesn't handle that :unamused:

Comment (tboerger) [reply Makefile:12]: Just keep it like it is. The `v` replacement can even be done with plain bash ;)

**Comment by strk** [comment]
Building the binary in the root is done by #247, please review that one too (and please someone restart drone)

**Comment by metalmatze** [comment]
Hey @bkcsoft, can you rebase this PR. There's one conflict. Thanks! 😊

**Comment by lunny** [comment]
Who can test if .VERSION be updated when `make build`

**Comment by thibaultmeyer** [comment]
> When someone merges this, please Squash Merge it. Don't need 8 commits in the history for this simple fix 

@bkcsoft if you have admin right on repo, you can disable "merge commit" to force each PRs to be merged with squash or rebase options.

image

**Comment by bkcsoft** [comment]
@0xbaadf00d for the most part I'm fine with merges 😉

**Comment by strk** [comment]
Time to squash-rebase and force-push @bkcsoft - the build-in-root was merged, so we can now see if `go build` and `make` both produce a usable binary with and w/out autogenerated version

**Comment by lunny** [comment]
@strk That's not the emphasis. The thing is `templates/.VERSION` has not been updated, so that you cannot run `git web` normallly. But maybe this could be resolved after #314 merged. It removed `templates/.VERSION`

**Comment by bkcsoft** [comment]
@lunny only possibility I see is to always re-generate on _every_ build (even when not necessary)

**Comment by bkcsoft** [comment]
And yes, I'd prefer to wait for #314 so I don't have to do it at all 😆

**Comment by bkcsoft** [comment]
Rebased on current master, fixed the issue with `.VERSION` not being built...

@lunny please re-test when you have time 🙂

File: Makefile
Comment (lunny) [inline Makefile]: maybe we should change this to `echo $(VERSION) > $@`?

Comment (bkcsoft) [reply Makefile]: no, it should be without `\n`, otherwise the version-check fails...

Comment (lunny) [reply Makefile]: I run `make build` on MacOS 10.12.1
```
code.gitea.io/gitea/cmd
code.gitea.io/gitea
echo -n 0.9.99+418-g399b0f7 > templates/.VERSION
➜  gitea git:(91-auto-build-version) ✗ more templates/.VERSION
-n 0.9.99+418-g399b0f7
➜  gitea git:(91-auto-build-version) ✗
```

Comment (tboerger) [reply Makefile]: IMHO we should anyway trim the result within the version check

Comment (bkcsoft) [reply Makefile]: the HELL apple, stop fucking around with my flags! 😆 

I'll trim it instead... maybe I should drop it? I think that @tboerger PR for removing it is merged no?

Comment (lunny) [reply Makefile]: Yes, you can drop it and wait until he's PR merged.

Comment (lunny) [reply Makefile]: @bkcsoft, just drop it. I think this could be merged.

**Comment by bkcsoft** [comment]
@lunny Done :)

**Review by lunny — APPROVED** [review]

**Pull request merged**
